third_party/move/move-vm/types/src/values/values_impl.rs

```diff
@@ -662,11 +690,102 @@ impl Container {
 }
 
 impl IndexedRef {
+    /*
+     * Runtime enum-variant tag guard for IndexedRef
+     *
+     * Motivation
+     * ----------
+     * Move’s bytecode verifier enforces reference-safety (no aliasing violations, no destructive
+     * updates through aliases, etc.). However, if a verifier bug allows an enum value to be overwritten
+     * via a mutable alias while an immutable field reference to that enum is still live, a classic “stale field reference”
+     * can arise: the reference was created when the enum had variant A, but later the enum is rewritten to variant B, so reading
+     * the field at payload index 1 would be a type confusion.
+     *
+     * Idea
+     * ----
+     * Whenever a field reference is created via `borrow_variant_field`, the VM captures the
+     * current enum variant tag (a u16 stored at field 0 in the Struct backing the enum) and
+     * stores it in the `IndexedRef` (as `self.tag = Some(tag)`). On every later use of that
```

Comment:
1. What is the performance penalty of this check? For example, what is the impact on the decibel benchmarks (or other standard benchmarks we have)?
2. Does a "write" constitute a use as well, or is it only "reads"?